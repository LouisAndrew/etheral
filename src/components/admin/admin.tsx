import React, { useState } from 'react';
import { sha256 } from 'js-sha256';

import Login from 'components/auth/login';
import { LoginProps } from 'components/auth/auth';

type Props = {
    db: firebase.firestore.Firestore;
};

const Admin: React.FC<Props> = ({ db }) => {
    const [isAuthenticated, setIsAuthenticated] = useState(false);

    const adminUserDbRef = db.collection('admin-user');

    const login = async ({ email, password }: LoginProps) => {
        // authenticate user first
        const adminUser = await adminUserDbRef
            .where('email', '==', email)
            .where('password', '==', sha256(password))
            .get();

        const adminUserExists = (await adminUser.size) > 0;

        if (await adminUserExists) {
            setIsAuthenticated(true);
        } else {
            console.log('unauthorized');
            console.log(sha256(password));
        }
    };

    return (
        <>
            {isAuthenticated ? (
                <div data-testid="dashboard">
                    <h2>Welcome to admin dashboard</h2>
                    <button
                        onClick={() => {
                            setIsAuthenticated(false);
                        }}
                    >
                        Log out
                    </button>
                </div>
            ) : (
                <Login login={login} submitValue="Login as admin" />
            )}
        </>
    );
};

export { Admin };
