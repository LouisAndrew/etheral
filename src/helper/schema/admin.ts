export type Admin = {
    email: string;
    password: string;
    invitedBy: string;
};
