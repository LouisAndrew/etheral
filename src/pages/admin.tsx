import React from 'react';
import { PageProps } from 'gatsby';

import Admin from 'components/admin';

const AdminPage = (props: PageProps) => <Admin />;

export default AdminPage;
