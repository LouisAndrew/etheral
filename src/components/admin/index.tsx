import React, { useEffect, useState } from 'react';
import firebase from 'gatsby-plugin-firebase';
import { Admin as AdminEl } from './admin';

const Admin = () => {
    const [db, setDb] = useState<firebase.firestore.Firestore | undefined>(
        undefined
    );

    useEffect(() => {
        setDb(firebase.firestore());
    }, []);

    return db ? <AdminEl db={db} /> : <></>;
};

export default Admin;
