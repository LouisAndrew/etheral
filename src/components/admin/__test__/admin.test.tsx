import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import { render, cleanup } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import { sha256 } from 'js-sha256';

import { Admin } from '../admin';

describe('Admin page', () => {
    const mockUserExist = {
        email: 'foo@example.com',
        pass: '123456',
    };

    const mockDb = ({
        collection: jest.fn(() => ({
            where: jest.fn((str: string, comp: string, email: string) => ({
                where: jest.fn((str: string, comp: string, pass: string) => ({
                    get: jest.fn(() => {
                        if (
                            email === mockUserExist.email &&
                            pass === sha256(mockUserExist.pass)
                        ) {
                            return {
                                size: 1,
                            };
                        } else {
                            return {
                                size: 0,
                            };
                        }
                    }),
                    str,
                    comp,
                })),
                str,
                comp,
            })),
        })),
    } as any) as firebase.firestore.Firestore;

    const el: React.ReactElement = <Admin db={mockDb} />;

    afterEach(cleanup);

    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(el, div);
    });

    it('should not give access to admin dashboard if the user authenticated is not on the admin-user database', () => {
        const { queryByTestId, getByLabelText, getByDisplayValue } = render(el);
        const adminEmailInput = getByLabelText('Email');
        const adminPassInput = getByLabelText('Password');
        const submitInput = getByDisplayValue('Login as admin');

        userEvent.type(adminEmailInput, 'nonexistent@email.com');
        userEvent.type(adminPassInput, 'wrongpassword');

        userEvent.click(submitInput);

        setTimeout(() => {
            const adminDashboard = queryByTestId('dashboard');
            if (adminDashboard) {
                fail();
            } else {
                expect(adminDashboard).toBeNull();
            }
        }, 500);
    });

    it('should give access to admin dashboard if the user authenticated is on the admin-user database', () => {
        const { queryByTestId, getByLabelText, getByDisplayValue } = render(el);
        const adminEmailInput = getByLabelText('Email');
        const adminPassInput = getByLabelText('Password');
        const submitInput = getByDisplayValue('Login as admin');

        userEvent.type(adminEmailInput, mockUserExist.email);
        userEvent.type(adminPassInput, mockUserExist.pass);

        userEvent.click(submitInput);

        setTimeout(() => {
            const adminDashboard = queryByTestId('dashboard');

            if (!adminDashboard) {
                fail();
            } else {
                expect(adminDashboard).toBeInTheDocument();
            }

            const logoutButton = getByDisplayValue('Log out');
            userEvent.click(logoutButton);

            expect(adminDashboard).not.toBeInTheDocument();
        }, 500);
    });

    it('matches snapshot', () => {
        const tree = renderer.create(el).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
