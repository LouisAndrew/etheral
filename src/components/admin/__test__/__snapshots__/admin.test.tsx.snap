// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Admin page matches snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <label
    htmlFor="login-email"
  >
    Email
  </label>
  <input
    id="login-email"
    name="loginEmail"
    type="email"
  />
  <label
    htmlFor="login-password"
  >
    Password
  </label>
  <input
    id="login-password"
    name="loginPassword"
    type="password"
  />
  <input
    type="submit"
    value="Login as admin"
  />
</form>
`;
